File: core/src/main/java/com/triangleleft/flashcards/ui/login/LoginViewActions.java
package com.triangleleft.flashcards.ui.login;

import com.triangleleft.flashcards.util.TextUtils;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public abstract class LoginViewActions {

    private final static Logger logger = LoggerFactory.getLogger(LoginViewActions.class.getSimpleName());

    public static ViewAction<LoginViewState> setLogin(final String login) {
        logger.debug("setLogin {}", login);
        return state -> state.asBuilder()
                .login(login)
                .loginButtonEnabled(isLoginButtonEnabled(login, state.password()))
                .build();
    }

    public static ViewAction<LoginViewState> setPassword(final String password) {
        logger.debug("setPassword {}", password);
        return state -> state.asBuilder()
                .password(password)
                .loginButtonEnabled(isLoginButtonEnabled(state.login(), password))
                .build();
    }

    private static boolean isLoginButtonEnabled(String login, String password) {
        logger.debug("isLoginButtonEnabled: {} {}", login, password);
        return TextUtils.hasText(login) && TextUtils.hasText(password);
    }

    public static ViewAction<LoginViewState> setRememberUser(boolean rememberUser) {
        return state -> state.asBuilder().shouldRememberUser(rememberUser).build();
    }

    public static ViewAction<LoginViewState> loginError() {
        return state -> state.asBuilder().page(LoginViewState.Page.CONTENT)
                .hasLoginError(true)
                .build();
    }

    public static ViewAction<LoginViewState> passwordError() {
        return state -> state.asBuilder().page(LoginViewState.Page.CONTENT)
                .hasPasswordError(true).build();
    }

    public static ViewAction<LoginViewState> genericError(final String genericError) {
        return state -> state.asBuilder().page(LoginViewState.Page.CONTENT).genericError(genericError)
                .build();
    }

    public static ViewAction<LoginViewState> success() {
        return state -> state.asBuilder().page(LoginViewState.Page.SUCCESS).build();
    }

    public static ViewAction<LoginViewState> progress() {
        return state -> state.asBuilder().page(LoginViewState.Page.PROGRESS).build();
    }
}

File: core/src/main/kotlin/com/triangleleft/flashcards/ui/login/LoginViewActions.kt
package com.triangleleft.flashcards.ui.login

import com.triangleleft.flashcards.ui.ViewAction
import com.triangleleft.flashcards.util.TextUtils
import org.slf4j.LoggerFactory

object LoginViewActions {

    private val logger = LoggerFactory.getLogger(LoginViewActions::class.java.simpleName)

    fun setLogin(login: String): ViewAction<LoginViewState> {
        logger.debug("setLogin {}", login)
        return ViewAction { it.copy(login = login, loginButtonEnabled = isLoginButtonEnabled(login, it.password)) }
    }

    fun setPassword(password: String): ViewAction<LoginViewState> {
        logger.debug("setPassword {}", password)
        return ViewAction { it.copy(password = password, loginButtonEnabled = isLoginButtonEnabled(it.login, password)) }
    }

    private fun isLoginButtonEnabled(login: String, password: String): Boolean {
        logger.debug("isLoginButtonEnabled: {} {}", login, password)
        return TextUtils.hasText(login) && TextUtils.hasText(password)
    }

    fun setRememberUser(rememberUser: Boolean): ViewAction<LoginViewState> {
        return ViewAction { it.copy(shouldRememberUser = rememberUser) }
    }

    fun loginError(): ViewAction<LoginViewState> {
        return ViewAction { it.copy(page = LoginViewState.Page.CONTENT, hasLoginError = true) }
    }

    fun passwordError(): ViewAction<LoginViewState> {
        return ViewAction { it.copy(page = LoginViewState.Page.CONTENT, hasPasswordError = true) }
    }

    fun genericError(genericError: String): ViewAction<LoginViewState> {
        return ViewAction { it.copy(page = LoginViewState.Page.CONTENT, genericError = genericError) }
    }

    fun success(): ViewAction<LoginViewState> {
        return ViewAction { it.copy(page = LoginViewState.Page.SUCCESS) }
    }

    fun progress(): ViewAction<LoginViewState> {
        return ViewAction { it.copy(page = LoginViewState.Page.PROGRESS) }
    }
}
